File: exercises/domain-5/block-5.2/solution/solution.ts
import { setTimeout as sleep } from "timers/promises"

type ErrorClass = "retry" | "escalate" | "degrade"

interface EscalationRecord {
  step: string
  error: string
  timestamp: string
  resolution: string
}

class EscalationManager {
  escalationLog: EscalationRecord[] = []
  private failureCounts = new Map<string, number>()
  private readonly RETRY_LIMIT = 3

  classifyError(error: Error): ErrorClass {
    const msg = error.message.toLowerCase()
    // Escalate immediately
    if (msg.includes("permission") || msg.includes("eperm")) return "escalate"
    if (msg.includes("safety") || msg.includes("violation")) return "escalate"
    if (msg.includes("corrupt") || msg.includes("invalid data")) return "escalate"
    // Retry
    if (msg.includes("network") || msg.includes("econnrefused")) return "retry"
    if (msg.includes("timeout") || msg.includes("rate limit") || msg.includes("429")) return "retry"
    // Degrade (unknown/optional)
    return "degrade"
  }

  async handleWithEscalation<T>(
    stepName: string,
    fn: () => Promise<T>,
    options = { required: true }
  ): Promise<T | null> {
    let attempts = 0

    while (true) {
      try {
        const result = await fn()
        this.failureCounts.set(stepName, 0)
        return result
      } catch (err) {
        attempts++
        const error = err instanceof Error ? err : new Error(String(err))
        const classification = options.required ? this.classifyError(error) : "degrade"

        const count = (this.failureCounts.get(stepName) ?? 0) + 1
        this.failureCounts.set(stepName, count)

        if (classification === "escalate" || count >= this.RETRY_LIMIT) {
          const resolution = count >= this.RETRY_LIMIT
            ? `Escalated after ${count} failures`
            : `Escalated: ${classification} error`
          this.escalationLog.push({
            step: stepName,
            error: error.message,
            timestamp: new Date().toISOString(),
            resolution,
          })
          console.log(`  🚨 ESCALATED: ${stepName} — ${error.message}`)
          return null
        }

        if (classification === "degrade") {
          this.escalationLog.push({
            step: stepName, error: error.message,
            timestamp: new Date().toISOString(), resolution: "Degraded (optional step)"
          })
          console.log(`  ⚠️ DEGRADED: ${stepName} — returning null`)
          return null
        }

        // Retry
        console.log(`  🔄 Retry ${attempts} for ${stepName}: ${error.message}`)
        await sleep(100 * attempts)
      }
    }
  }
}

async function main() {
  const manager = new EscalationManager()
  let attempts = 0

  const result1 = await manager.handleWithEscalation("fetch-data", async () => {
    attempts++
    if (attempts < 3) throw new Error("ECONNREFUSED: network error")
    return { data: "success" }
  })
  console.log("Step 1:", result1)

  const result2 = await manager.handleWithEscalation("write-db", async () => {
    throw new Error("EPERM: permission denied")
  })
  console.log("Step 2:", result2)

  const result3 = await manager.handleWithEscalation("generate-chart", async () => {
    throw new Error("Chart service unavailable")
  }, { required: false })
  console.log("Step 3:", result3)

  console.log("\nEscalation log:", manager.escalationLog)
}

main().catch(console.error)

File: exercises/domain-5/block-5.2/solution/solution.py
import asyncio
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import TypeVar, Callable, Awaitable

T = TypeVar("T")


@dataclass
class EscalationRecord:
    step: str
    error: str
    timestamp: str
    resolution: str


class EscalationManager:
    def __init__(self):
        self.escalation_log: list[EscalationRecord] = []
        self._failure_counts: dict[str, int] = {}
        self.RETRY_LIMIT = 3

    def classify_error(self, error: Exception) -> str:
        msg = str(error).lower()
        if "permission" in msg or "eperm" in msg: return "escalate"
        if "safety" in msg or "violation" in msg: return "escalate"
        if "corrupt" in msg: return "escalate"
        if "network" in msg or "econnrefused" in msg: return "retry"
        if "timeout" in msg or "rate limit" in msg: return "retry"
        return "degrade"

    async def handle_with_escalation(self, step_name: str, fn: Callable, required: bool = True):
        attempts = 0
        while True:
            try:
                result = await fn()
                self._failure_counts[step_name] = 0
                return result
            except Exception as err:
                attempts += 1
                classification = self.classify_error(err) if required else "degrade"
                count = self._failure_counts.get(step_name, 0) + 1
                self._failure_counts[step_name] = count

                if classification == "escalate" or count >= self.RETRY_LIMIT:
                    resolution = f"Escalated after {count} failures" if count >= self.RETRY_LIMIT else "Escalated: immediate"
                    self.escalation_log.append(EscalationRecord(
                        step=step_name, error=str(err),
                        timestamp=datetime.now(timezone.utc).isoformat(), resolution=resolution
                    ))
                    print(f"  🚨 ESCALATED: {step_name}")
                    return None

                if classification == "degrade":
                    self.escalation_log.append(EscalationRecord(
                        step=step_name, error=str(err),
                        timestamp=datetime.now(timezone.utc).isoformat(), resolution="Degraded"
                    ))
                    print(f"  ⚠️ DEGRADED: {step_name}")
                    return None

                print(f"  🔄 Retry {attempts} for {step_name}: {err}")
                await asyncio.sleep(0.1 * attempts)


async def main():
    manager = EscalationManager()
    attempts = 0

    async def fetch_data():
        nonlocal attempts
        attempts += 1
        if attempts < 3:
            raise ConnectionError("ECONNREFUSED: network error")
        return {"data": "success"}

    print("Step 1:", await manager.handle_with_escalation("fetch-data", fetch_data))
    print("Step 2:", await manager.handle_with_escalation("write-db", lambda: (_ for _ in ()).throw(PermissionError("EPERM: permission denied"))))
    print("Step 3:", await manager.handle_with_escalation("generate-chart", lambda: (_ for _ in ()).throw(RuntimeError("Chart unavailable")), required=False))
    print("\nEscalation log:", manager.escalation_log)


asyncio.run(main())
